### DISCUSSION TITLE
Curious about similarities between Mojo and Terra

### DISCUSSION THREAD

**Discussion by cjohnson318** [modular/modular#502] [Q&A]
I remember [Terra](https://terralang.org/) coming out about ten years ago. It looked interesting, but I never used it, as I was more focused on Python, Swift and ObjectiveC at the time. Basically, it's Lua, but with a typed "terra" functions for low level, system-programming. I reminds me of Mojo, which is mostly Python syntax, with typed "fn" functions for greater speed and control. The obvious difference is that Mojo is designed with parallelism in mind, and it embraces the already huge Python community.

Does anyone have thoughts on why something as slick as Terra never gained traction, and the best ways to line up Mojo for success?

**Comment by drunkwcodes** [comment]
I used lua to write [minetest](https://github.com/minetest/minetest) game plugins once. It's certainly slick.
I think system programming is for minority of developers. And lua users are not many as well.

But mojo is different. Python developers are the majority.
And the llvm and compiler tech stack is the crucial part to attract me as a python dev.

When mojo has outstanding performance, and decent developer experience,
it will gain most of python dev's eyeball. 
Suddenly it will be full of vivid project ideas around mojo.

If mojo can defeat go in terms of parallelism, rust in terms of stability, julia in terms of affinity,
it will reach unprecedented availibility for all of the developers.

I do not help building the compiler, but
I'm working on building `muse`, a mojo project/package manager based on `pdm`.
And I will make a code formatter `tube` for mojo.
It's all about developer experience. Please contact me if you want to join me: `drunkwcodes@gmail.com`.

**Reply by cjohnson318** [reply]
Both projects offer the ability to opt-in to more performant, lower level patterns, from a familiar scripting language. I think this is a compelling and powerful idea. I think you're right that Mojo's basis in Python should do more to drive adoption. I also think that the focus on ease of parallelism and safety are important features of Mojo, over Terra. I hope that that it is enough! I am very excited about Mojo.

Your projects `muse` and `tube` sound promising!

**Comment by cjohnson318** [comment]
P.S. I saw this blog post about writing a package manager and I thought you might like it: https://antonz.org/writing-package-manager/

**Reply by drunkwcodes** [reply]
It is such a good reference of designing package manager.
